Mathlib/Algebra/Lie/OfAssociative.lean

```diff
@@ -252,6 +252,63 @@ theorem LieSubalgebra.toEndomorphism_mk (K : LieSubalgebra R L) {x : L} (hx : x
   rfl
 #align lie_subalgebra.to_endomorphism_mk LieSubalgebra.toEndomorphism_mk
 
+section
+
+open BigOperators LieAlgebra LieModule
+
+lemma LieSubmodule.coe_toEndomorphism (N : LieSubmodule R L M) (x : L) (y : N) :
+    (toEndomorphism R L N x y : M) = toEndomorphism R L M x y := rfl
+
+lemma LieSubmodule.coe_toEndomorphism_pow (N : LieSubmodule R L M) (x : L) (y : N) (n : ℕ) :
+    ((toEndomorphism R L N x ^ n) y : M) = (toEndomorphism R L M x ^ n) y := by
+  induction n generalizing y with
+  | zero => rfl
+  | succ n ih => simp only [pow_succ', LinearMap.mul_apply, ih, LieSubmodule.coe_toEndomorphism]
+
+lemma LieSubalgebra.coe_ad (H : LieSubalgebra R L) (x y : H) :
+    (ad R H x y : L) = ad R L x y := rfl
+
+lemma LieSubalgebra.coe_ad_pow (H : LieSubalgebra R L) (x y : H) (n : ℕ) :
+    ((ad R H x ^ n) y : L) = (ad R L x ^ n) y := by
+  induction n generalizing y with
+  | zero => rfl
+  | succ n ih => simp only [pow_succ', LinearMap.mul_apply, ih, LieSubalgebra.coe_ad]
+
+variable {L M}
+
+lemma LieModule.toEndomorphism_lie (x y : L) (z : M) :
+    (toEndomorphism R L M x) ⁅y, z⁆ = ⁅ad R L x y, z⁆ + ⁅y, toEndomorphism R L M x z⁆ := by
+  simp
+
+lemma LieAlgebra.ad_lie (x y z : L) :
+    (ad R L x) ⁅y, z⁆ = ⁅ad R L x y, z⁆ + ⁅y, ad R L x z⁆ :=
+  toEndomorphism_lie _ x y z
+
+open Finset in
+lemma LieModule.toEndomorphism_pow_lie (x y : L) (z : M) (n : ℕ) :
+    ((toEndomorphism R L M x) ^ n) ⁅y, z⁆ =
+      ∑ ij in antidiagonal n, n.choose ij.1 •
+        ⁅((ad R L x) ^ ij.1) y, ((toEndomorphism R L M x) ^ ij.2) z⁆ := by
+  induction n with
+  | zero => simp
+  | succ n ih =>
+    rw [Finset.sum_antidiagonal_choose_succ_nsmul
+      (fun i j ↦ ⁅((ad R L x) ^ i) y, ((toEndomorphism R L M x) ^ j) z⁆) n]
+    simp only [pow_succ, LinearMap.mul_apply, ih, map_sum, map_nsmul,
+      toEndomorphism_lie, nsmul_add, sum_add_distrib]
+    convert add_comm _ _ using 4
+    rename_i ij hij
+    rw [mem_antidiagonal, add_comm] at hij
+    exact Nat.choose_symm_of_eq_add hij.symm
```

Comment:
I claim this is slightly less brittle / more maintainable:
```suggestion
    rw [add_comm]
    refine congr_arg₂ _ rfl (sum_congr rfl fun ⟨i, j⟩ hij ↦ ?_)
    replace hij : n = j + i := by rwa [mem_antidiagonal, add_comm, eq_comm] at hij
    rw [Nat.choose_symm_of_eq_add hij]
```
(though feel free to disagree)

Reply:
Thanks, I combined this with some other changes. Hopefully you like the improvement.